designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,331 @@
+
+These debug configurations are authored in a json file. The following Toolkit assistance is provided:
+
+-   autocompletion with descriptions is provided for `aws-sam` related fields
+    -   There is no autocompletion available for specific values in a configuration. For example, if a user types in the location of a SAM Template file, there is no filesystem-based autocompletion. The Debug Configuration validates the configuration and notifies of errant values when it is run.
+-   snippets to produce typical (or starter) `aws-sam` debug configurations
+-   when no launch.json file is present in a workspace, VS Code exposes functionality that allows users to request auto-generated Debug Configurations. In this situation, the toolkit generates an `aws-sam` Debug Configuration for all `AWS::Serverless::Function` resources detected within all SAM Templates located in the workspace.
+
+Example Debug Configuration entries can be found in the [Appendix](#sample-debug-configurations)
+
+Standalone Lambda function handlers are not supported through Debug Configurations.
+
+### <a id="codelenses"></a> CodeLenses
+
+#### CodeLenses in SAM Template files
+
+CodeLenses are added to SAM Template files that serve as shortcuts to any [debug configurations](#debug-configurations) defined in the workspace.
+
+Every Debug Configuration that references a SAM Template and resource pairing will produce a CodeLens above that resource. When clicked, these CodeLenses launch the corresponding debug session, as if the user selected that debug configuration and pressed Debug from VS Code's Debug view.
+
+An additional CodeLens is placed above every template resource of type `AWS::Serverless::Function` called "Add Debug Configuration". Clicking this CodeLens produces a pre-filled `aws-sam` Debug Configuration in `launch.json` capable of performing a [local invoke](#sam-template-resource-local) of the associated resource.
+
+#### CodeLenses in Code files
+
+CodeLenses in code files provide support for debugging standalone Lambda function handlers.
+
+The following CodeLenses appear over any function that appears to be an eligible Lambda handler: Run Locally, Debug Locally, Configure.
+
+The Configure CodeLens allows users to customize how the function handler is locally debugged. Configuration is optional, using sensible defaults when necessary. The configurable aspects are listed [above](#debug-config). Each handler is configured separately, and all handler configurations are stored in the workspace at `.aws/lambda-handlers.json` ([file structure](#code-file-codelens-config)). Clicking the Configure CodeLens opens this file, adds a configuration entry for the corresponding handler if necessary, and places the cursor at the handler's entry. Users have autocompletion support in this file. A rich UI for configuration is not considered at this time, but the door remains open to adding a visual editor in the future based on user feedback.
```

Comment:
Will these configurations be transferable to the `launch.json` file? What happens if a user clicks a codelens on a function that already has a debug config in the `launch.json` file? Will we be able to tie these to VS Code's refactoring functionality, particularly if someone adds an elaborate configuration but later decides to change the handler's name?

Reply:
The Debug Configuration (launch.json) structures are not the same as these CodeLens based configuration structures, however we should aim to make them as similar as possible, both for familiarity when using them, and for code reuse in developing this feature. The Appendix has samples of both, where I've tried to reuse common shapes.

The two systems are completely separate. If you click a CodeLens, you can think of it like you are performing a one-off execution. If you use the Debug pane, you are running the debug configuration that is defined.

The refactoring concept is interesting, and could be a future value add, if its possible. Configuration entries aren't destroyed, so users would have to rename the key in the configuration file (similar to the current configuration experience).

Reply:
> The Debug Configuration (launch.json) structures are not the same as these CodeLens based configuration structures

Is there a reason we can't make them the same? Looks like it's only missing `name`, `type`, `request`, and `samTemplate`; `name` is arbitrary, `type`/`request` are static, and we may still want to add the `samTemplate`'s `parameters` tag to whatever temporary SAM template we're using. Doing this might make it a bit easier for a user to visualize what transferring these to a more concrete debugger would look like.